### PR TITLE
Multiple pools support for AsyncLoader

### PR DESCRIPTION
**Pull request by serxa**
AsyncLoader can now execute jobs in multiple thread pools. This allows controlling the concurrency of foreground and background tasks. Changes:
1. Notion of priority is now moved from job to pool. Job is instead assigned to a pool, which controls all scheduling aspects of jobs (concurrency and priority).
2. Job prioritization can now not only increase priority but also boost max concurrency because it moves jobs from one pool into another (possibly bigger).
3. Fixed problem of partial priority inversion due to not full utilization (e.g. for async tables loading worker could start executing startup jobs if there are no more load jobs ready, but loading of other tables is not finished yet). Now new jobs in lower priority pools are not started unless all higher priority jobs are finished.

Follow-up for https://github.com/ClickHouse/ClickHouse/pull/48923.
Inspired by `turboMode` https://github.com/ClickHouse/ClickHouse/pull/49474 (although takes a different approach)

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit e4a653b41cf109c7e4a885b3763a8f68f3080d89 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49893/e4a653b41cf109c7e4a885b3763a8f68f3080d89/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by serxa** [comment]
After rework on of unit test became flaky:
```
[ RUN      ] AsyncLoader.DynamicPriorities
/home/ubuntu/ClickHouse2/src/Common/tests/gtest_async_loader.cpp:686: Failure
Expected equality of these values:
  schedule
    Which is: "A4C4D9E9F9G9B3H0"
  "A4C4E9D9F9G9B3H0"
[  FAILED  ] AsyncLoader.DynamicPriorities (1 ms)
```
Working on it...

File: src/Common/AsyncLoader.cpp [hunk 1]
```diff
@@ -531,95 +568,138 @@ void AsyncLoader::finish(std::unique_lock<std::mutex> & lock, const LoadJobPtr &
...
 
-void AsyncLoader::prioritize(const LoadJobPtr & job, ssize_t new_priority, std::unique_lock<std::mutex> & lock)
+void AsyncLoader::prioritize(const LoadJobPtr & job, size_t new_pool_id, std::unique_lock<std::mutex> & lock)
 {
     if (auto info = scheduled_jobs.find(job); info != scheduled_jobs.end())
     {
-        if (info->second.priority >= new_priority)
-            return; // Never lower priority
+        Pool & old_pool = pools[job->pool_id];
+        Pool & new_pool = pools[new_pool_id];
```
[the remark this answers is not in the document]
Comment (serxa) [reply src/Common/AsyncLoader.cpp:576]: I assume it to be a logical error. But you are right, a simple check will be useful, I'll add it.

[the remark this answers is not in the document]
Comment (serxa) [reply src/Common/AsyncLoader.cpp:578]: Done

**Review by antonio2368 — APPROVED** [review]

**Comment by serxa** [comment]
Unit test (msan):
```
[ RUN      ] AsyncLoader.CancelPendingDependency
src/Common/tests/gtest_async_loader.cpp:308: Failure
Expected equality of these values:
  job1->status()
    Which is: 4-byte object <01-00 00-00>
  LoadStatus::CANCELED
    Which is: 4-byte object <03-00 00-00>
[  FAILED  ] AsyncLoader.CancelPendingDependency (0 ms)
```
Looks related... Checking...

**Pull request merged**